File: app/src/main/java/fnsb/macstransit/RouteMatch/Route.java
package fnsb.macstransit.RouteMatch;

import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.UiThread;

import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Polyline;
import com.google.android.gms.maps.model.PolylineOptions;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;

import fnsb.macstransit.Activities.MapsActivity;

/**
 * Created by Spud on 2019-10-12 for the project: MACS Transit.
 * <p>
 * For the license, view the file titled LICENSE at the root of the project.
 *
 * @version 2.6.
 * @since Beta 3.
 */
public class Route {

	/**
	 * Fallback empty route array used as a way to circumvent potential exceptions.
	 */
	public static final Route[] EMPTY_ROUTE = new Route[0];

	/**
	 * The name of the route.
	 */
	public final String routeName;

	/**
	 * The name of the route formatted to be parsed as a URL.
	 */
	public final String urlFormattedName; // TODO Add regex check

	/**
	 * The color of the route.
	 * This is optional, as there is a high chance that the route does not have one.
	 * <p>
	 * This is an int instead of a Color object as android stores its colors as an integer.
	 */
	public int color;

	/**
	 * The array of stops for this route.
	 * This may be empty / null if the route has not been initialized,
	 * and the stops haven't been loaded.
	 */
	@Nullable
	public Stop[] stops;

	/**
	 * Whether or not the route is enabled or disabled (to be shown or hidden).
	 * Default is false (disabled).
	 */
	public boolean enabled = false;

	/**
	 * The array of LatLng coordinates that will be used to create the polyline (if enabled).
	 * This is private as we don't want this variable to be set outside the class.
	 */
	private LatLng[] polyLineCoordinates;

	/**
	 * The array of shared stops for this route.
	 * This may be empty/ null if the route has not been initialized,
	 * and the the shared stops haven't been loaded, or if there are no shared stops for the route.
	 * <p>
	 * The reason why this is private access is because there is a special way this should be set within the route class.
	 */
	private SharedStop[] sharedStops;

	/**
	 * The polyline that corresponds to this route.
	 * The reason why this is private access is because there is a special way this should be set within the route class.
	 */
	private Polyline polyline;

	/**
	 * Constructor for the route. The name of the route is the only thing that is required.
	 * Be sure that the provided route name does <b>NOT</b> contain any whitespace characters!
	 *
	 * @param routeName The name of the route. Be sure this does <b>NOT</b>
	 *                  contain any whitespace characters!
	 * @throws UnsupportedEncodingException Thrown if the route name cannot be formatted to a URL.
	 */
	public Route(@NonNull String routeName) throws UnsupportedEncodingException {
		this(routeName, 0);
	}

	/**
	 * Constructor for the route. The name of the route is the only thing that is required.
	 * Be sure that the provided route name does <b>NOT</b> contain any whitespace characters!
	 *
	 * @param routeName The name of the route. Be sure this does <b>NOT</b>
	 *                  contain any whitespace characters!
	 * @param color     The route's color. This is optional,
	 *                  and of the color is non-existent simply use the
	 *                  {@code Route(String routeName)} constructor.
	 * @throws UnsupportedEncodingException Thrown if the route name cannot be formatted to a URL.
	 */
	public Route(String routeName, int color) throws UnsupportedEncodingException {

		// Set the route name.
		this.routeName = routeName;

		// Parse and set the url from the name.
		// The URL encoding that the RouteMatch API uses is slightly different than that provided by the URL encoder
		this.urlFormattedName = java.util.regex.Pattern.compile("\\+").matcher(java.net.URLEncoder.
				encode(routeName, "UTF-8")).replaceAll("%20");

		// Set the route color.
		this.color = color;
	}

	/**
	 * Generates the array of routes from the master schedule json array.
	 * If a null array is provided then an empty array will be returned.
	 *
	 * @param masterSchedule The json array containing the master schedule containing all the routes.
	 * @return The routes derived from the master schedule.
	 */
	@NonNull
	public static Route[] generateRoutes(JSONArray masterSchedule) {

		// Make sure the master schedule is not null. If it is, return an empty route array.
		if (masterSchedule == null) {
			Log.w("generateRoutes", "Master schedule is null!");
			return Route.EMPTY_ROUTE;
		}

		// Create an array to store all the generated routes.
		int count = masterSchedule.length();
		Route[] potentialRoutes = new Route[count];
		int routeCount = 0;

		// Iterate though each route in the master schedule.
		for (int index = 0; index < count; index++) {
			Log.d("generateRoutes", String.format("Parsing route %d/%d", index + 1, count));

			// Try to get the route data from the array.
			// If there's an issue parsing the data simply continue to the next iteration of the loop.
			JSONObject routeData;
			try {
				routeData = masterSchedule.getJSONObject(index);
			} catch (JSONException e) {
				Log.w("generateRoutes", "Issue retrieving the route data", e);
				continue;
			}

			// Try to create the route using the route data obtained above.
			// If there was a route exception thrown simply log it.
			try {
				Route route = Route.generateRoute(routeData);
				potentialRoutes[routeCount] = route;
				routeCount++;
			} catch (RouteException | UnsupportedEncodingException e) {
				Log.w("generateRoutes", "Issue creating route from route data", e);
			}
		}

		// Down size our potential routes array to fit the actual number of routes.
		Route[] routes = new Route[routeCount];
		System.arraycopy(potentialRoutes, 0, routes, 0, routeCount);
		return routes;
	}

	/**
	 * Creates a new route object from the provided json object.
	 * If the json object is null then a RouteException will be thrown.
	 *
	 * @param jsonObject The json object contain the data to create a new route object.
	 * @return The newly created route object.
	 * @throws RouteException               Thrown if the json object is null, or if the route name is unable to be parsed.
	 * @throws UnsupportedEncodingException Thrown if the route name cannot be formatted to a URL.
	 */
	@NonNull
	private static Route generateRoute(JSONObject jsonObject) throws RouteException, UnsupportedEncodingException {

		// Make sure the provided json object is not null.
		if (jsonObject == null) {
			throw new RouteException("Json object cannot be null!");
		}

		// First, parse the name.
		String name;
		try {
			name = jsonObject.getString("routeId");
		} catch (JSONException e) {
			throw new RouteException("Unable to get route name from JSON", e.getCause());
		}

		// Now try to parse the route and route color.
		Route route;
		try {
			String colorName = jsonObject.getString("routeColor");
			int color = android.graphics.Color.parseColor(colorName);

			route = new Route(name, color);

		} catch (JSONException | IllegalArgumentException e) {
			Log.w("generateRoute", "Unable to parse color");

			// Since there was an issue parsing the color, and we have the name at this point...
			// Simply create the route without a color.
			route = new Route(name);
		}

		// Return the newly created route.
		return route;
	}

	/**
	 * Iterates though all the routes in MapsActivity.allRoutes
	 * and enables those that have been favorited (as determined by being in the favoritedRoutes array).
	 * <p>
	 * This should only be run once.
	 *
	 * @param favoritedRoutes The selected routes to be enabled from MapsActivity.allRoutes.
	 */
	public static void enableFavoriteRoutes(Route[] favoritedRoutes) {

		// Make sure there are routes to iterate over.
		if (MapsActivity.allRoutes == null || favoritedRoutes == null) {
			return;
		}

		// Iterate through all the routes that will be used in the activity.
		for (Route allRoute : MapsActivity.allRoutes) {

			// Iterate though the favorite routes.
			for (Route favoritedRoute : favoritedRoutes) {

				// Make sure the favorited route and the comparison route aren't null.
				if (favoritedRoute != null && allRoute != null) {

					// If the route name matches the favorited route name, enable it.
					if (allRoute.routeName.equals(favoritedRoute.routeName)) {
						allRoute.enabled = true;
						break;
					}
				} else {

					// If either of the routes were null, log it.
					Log.w("enableFavoriteRoutes", "Route entry in array is null");
				}
			}
		}

		// Set the selectedFavorites variable to be true as to not run again.
		MapsActivity.selectedFavorites = true;
	}

	/**
	 * Starts an asynchronous process to fetch all the stops for the route from the RouteMatch API.
	 */
	public void loadStops() {

		// Get all the stops for the route from the RouteMatch object.
		MapsActivity.routeMatch.callAllStops(this, result -> {

			// Get the data from all the stops and store it in a JSONArray.
			JSONArray data = RouteMatch.parseData(result);

			// Load in all the potential stops for the route.
			// The reason why this is considered potential stops is because at this stage duplicate
			// stops have not yet been handled.
			Stop[] potentialStops = Stop.generateStops(data, this);

			// At this point duplicate stops have now been handled and removed.
			this.stops = Stop.validateGeneratedStops(potentialStops);

		}, error -> Log.w("loadStops", "Unable to get stops from RouteMatch server", error), this);
	}

	/**
	 * Loads the polyline coordinates for the route object by retrieving the array from the RouteMatch server.
	 * This method will either set the polyline coordinates for the route,
	 * or will return early if the route match object is null.
	 */
	public void loadPolyLineCoordinates() {

		// Make sure the RouteMatch object exists.
		if (MapsActivity.routeMatch == null) {
			Log.w("loadPolyLineCoordinates", "RouteMatch object is null!");
			return;
		}

		// Get the land route from the routematch API using an asynchronous process.
		MapsActivity.routeMatch.callLandRoute(this, response -> {

			try {
				// Get the land route data array from the land route object.
				JSONArray landRouteData = RouteMatch.parseData(response);

				// Get the land route points object from the land route data array.
				JSONObject landRoutePoints = landRouteData.getJSONObject(0);

				// Get the land route points array from the land route points object.
				JSONArray landRoutePointsArray = landRoutePoints.getJSONArray("points");

				// Get the number of points in the array.
				int count = landRoutePointsArray.length();

				// Create a new LatLng array to store all the coordinates.
				LatLng[] coordinates = new LatLng[count];

				// Initialize the array of coordinates by iterating through the land route points array.
				for (int i = 0; i < count; i++) {

					// Get the land route point object from the land route points array.
					JSONObject landRoutePoint = landRoutePointsArray.getJSONObject(i);

					// Get the latitude and longitude from the land route point.
					double latitude = landRoutePoint.getDouble("latitude"),
							longitude = landRoutePoint.getDouble("longitude");

					// Create a new LatLng object using the latitude and longitude.
					LatLng latLng = new LatLng(latitude, longitude);

					// Add the newly created LatLng object to the LatLng array.
					coordinates[i] = latLng;
				}

				// Set the polyline coordinates array to the finished LatLng array.
				this.polyLineCoordinates = coordinates;

			} catch (JSONException exception) {
				Log.e("loadPolyLineCoordinates", "Error parsing json", exception);
			}
		}, error -> Log.w("loadPolyLineCoordinates",
				"Unable to get polyline from routematch server", error), this);

	}

	/**
	 * Returns the polyline that corresponds to the route.
	 *
	 * @return The polyline that corresponds to the route.
	 */
	public Polyline getPolyline() {
		return this.polyline;
	}

	/**
	 * Creates and sets the polyline for the route.
	 * If there are no polyline coordinates for the route then this simply returns early and does not create the polyline.
	 */
	@UiThread
	public void createPolyline() {
		Log.v("createPolyline", "Creating route polyline");

		// Make sure the polyline coordinates is not null or 0. If it is then return early.
		if (this.polyLineCoordinates == null || this.polyLineCoordinates.length == 0) {
			Log.w("createPolyline", "There are no polyline coordinates to work with!");
			return;
		}

		// Make sure the map isn't null.
		if (MapsActivity.map == null) {
			Log.w("createPolyline", "Map is not yet ready!");
		}

		// Create new polyline options from the array of polyline coordinates stored for the route.
		PolylineOptions options = new PolylineOptions().add(this.polyLineCoordinates);

		// Make sure its not clickable.
		options.clickable(false);

		// Set the color of the polylines based on the route color.
		options.color(this.color);

		// Make sure the polyline starts out invisible.
		options.visible(false);

		// Add the polyline to the map, and set it for the object.
		this.polyline = MapsActivity.map.addPolyline(options);
	}

	/**
	 * Adds the shared to the routes shared stop array.
	 *
	 * @param sharedStop The shared stop to add to the route.
	 */
	public void addSharedStop(SharedStop sharedStop) {

		// Create a new shared stop array that will contain our current shared stop array + the new shared stop.
		SharedStop[] newSharedStops;

		if (this.sharedStops == null) {

			// If there was no shared stop array before then simply set the array to just contain our shared stop.
			newSharedStops = new SharedStop[]{sharedStop};
		} else {

			// Since our current array of shared stops has content
			// simply insert our shared stop to the array and copy the rest using System.arraycopy.
			newSharedStops = new SharedStop[this.sharedStops.length + 1];
			newSharedStops[0] = sharedStop;
			System.arraycopy(this.sharedStops, 0, newSharedStops, 1, this.sharedStops.length);
		}

		// Set the routes shared stop array to the newly created shared stop array.
		this.sharedStops = newSharedStops;
	}

	/**
	 * Gets the shared stops for the route. This may be null if there are none.
	 *
	 * @return The shared stops for the route.
	 */
	@Nullable
	public SharedStop[] getSharedStops() {
		return this.sharedStops;
	}

	/**
	 * Gets the LatLng object array for the polyline coordinates for the route.
	 *
	 * @return The polyline coordinates for the route as a LatLng array.
	 */
	@Nullable
	public LatLng[] getPolyLineCoordinates() {
		return this.polyLineCoordinates;
	}

	/**
	 * Removes the routes polyline from the map, and sets it to null.
	 * This must be run on the UI thread.
	 */
	@UiThread
	public void removePolyline() {
		if (this.polyline != null) {
			this.polyline.remove();
			this.polyline = null;
		}
	}

	/**
	 * Exception class used for throwing any exception relating to Routes.
	 */
	public static class RouteException extends Exception {

		/**
		 * Constructor for a new exception with a (hopefully detailed) message.
		 *
		 * @param message The (ideally detailed) message for why this was thrown.
		 */
		public RouteException(String message) {
			super(message);
		}

		/**
		 * Constructor for a new exception with a (hopefully detailed) message, and a cause.
		 *
		 * @param message The (ideally detailed) message.
		 * @param cause   The cause for the exception. This may be null if the cause is undetermined.
		 */
		public RouteException(String message, @Nullable Throwable cause) {
			super(message, cause);
		}

		/**
		 * Constructor for a new exception with a cause.
		 *
		 * @param cause The cause for the exception. This may be null if the cause is undetermined.
		 */
		public RouteException(@Nullable Throwable cause) {
			super(cause);
		}
	}
}

File: app/src/main/java/fnsb/macstransit/routematch/Route.kt
package fnsb.macstransit.routematch

import android.graphics.Color
import android.os.Build
import android.os.Build.VERSION
import android.os.Parcel
import android.os.Parcelable
import android.util.Log
import androidx.annotation.UiThread
import com.android.volley.VolleyError
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Polyline
import com.google.maps.android.ktx.addPolyline
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.io.UnsupportedEncodingException
import java.net.URLEncoder
import java.util.regex.Pattern

/**
 * Created by Spud on 2019-10-12 for the project: MACS Transit.
 * For the license, view the file titled LICENSE at the root of the project.
 *
 * @version 3.1.
 * @since Beta 3.
 */
class Route: Parcelable {

	/**
	 * The name of the route.
	 */
	val name: String

	/**
	 * The color of the route.
	 */
	var color: Int = 0

	/**
	 * The name of the route formatted to be parsed as a URL.
	 */
	val urlFormattedName: String

	/**
	 * All the stops in the route.
	 */
	val stops: HashMap<String, Stop> = HashMap()

	/**
	 * All the shared stops in the route.
	 */
	val sharedStops: HashMap<String, SharedStop> = HashMap()

	/**
	 * Whether or not the route is enabled or disabled (to be shown or hidden).
	 * Default is false (disabled).
	 */
	@Transient
	var enabled = false

	/**
	 * The polyline that corresponds to this route.
	 */
	@Transient
	private var polyline: Polyline? = null

	/**
	 * Creates a new Route object using information from the provided parcel.
	 * @param parcel The parcel containing all the route information.
	 */
	constructor(parcel: Parcel) {
		Log.v("Route", "Reading from parcel")

		// Load the name, color, and formatted name from the parcel.
		name = parcel.readString()!!
		color = parcel.readInt()
		urlFormattedName = parcel.readString()!!

		// Load the array of stops from the parcel.
		if (VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
			val stopArray: Array<Stop>? = parcel.readParcelableArray(Stop::class.java.classLoader, Stop::class.java)
			if (stopArray != null) {
				for (stop in stopArray) {
					stops[stop.name] = stop
				}
			}
		} else {
			@Suppress("DEPRECATION") // Suppressed because the function is replaced in newer APIs
			val stopParcelableArray: Array<Parcelable>? = parcel.readParcelableArray(Stop::class.java.classLoader)
			if (stopParcelableArray != null) {
				for (stop in stopParcelableArray) {
					if (stop is Stop) {
						stops[stop.name] = stop
					}
				}
			}
		}

		// Load the array of shared stops from the parcel.
		if (VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
			val sharedStopParcelableArray: Array<SharedStop>? = parcel.readParcelableArray(SharedStop::class.java.classLoader, SharedStop::class.java)
			if (sharedStopParcelableArray != null) {
				for (sharedStop in sharedStopParcelableArray) {
					sharedStops[sharedStop.name] = sharedStop
				}
			}
		} else {
			@Suppress("DEPRECATION") // Suppressed because the function is replaced in newer APIs
			val sharedStopParcelableArray: Array<Parcelable>? = parcel.readParcelableArray(SharedStop::class.java.classLoader)
			if (sharedStopParcelableArray != null) {
				for (sharedStop in sharedStopParcelableArray) {
					if (sharedStop is SharedStop) {
						sharedStops[sharedStop.name] = sharedStop
					}
				}
			}
		}

	}

	/**
	 * Constructor for a new Route object with only a name.
	 *
	 * @param routeName The name of the route. This cannot contain any whitespace characters.
	 * @throws UnsupportedEncodingException Thrown if the route name cannot be formatted to a URL.
	 */
	@Throws(UnsupportedEncodingException::class)
	constructor(routeName: String) : this(routeName, 0)

	/**
	 * Creates a new Route object.
	 *
	 * @param name The name of the route. This cannot contain any whitespace characters.
	 * @param color The route color.
	 * @throws UnsupportedEncodingException Thrown if the route name contains whitespace.
	 */
	@Throws(UnsupportedEncodingException::class)
	constructor(name: String, color: Int) {

		// Make sure the route name does not contain any whitespace characters.
		if (name.contains(Regex("\\s"))) {
			throw UnsupportedEncodingException("Route name contains whitespace!")
		}

		// Set the route name and color.
		this.name = name
		this.color = color

		// Set the urlFormattedName from the name.
		urlFormattedName = Pattern.compile("\\+").matcher(URLEncoder.encode(this.name, "UTF-8"))
			.replaceAll("%20")
	}

	/**
	 * Creates and sets the polyline for the route.
	 *
	 * This must be run on the UI Thread.
	 *
	 * @param coordinates The coordinates of the polyline.
	 * @param map The map to add the polyline to.
	 */
	@UiThread
	fun createPolyline(coordinates: Array<LatLng>, map: GoogleMap) {
		Log.v("createPolyline", "Creating route polyline")

		// Add the polyline to the map with the following options:
		polyline = map.addPolyline {

			// Create new polyline options from the array of polyline coordinates stored for the route.
			add(*coordinates)

			// Make sure its not clickable.
			clickable(false)

			// Set the color of the polylines based on the route color.
			color(color)

			// Set the polyline visibility to whether or not the route is enabled.
			visible(enabled)
		}
	}

	/**
	 * Removes stops that have shared stops.
	 */
	fun purgeStops() {

		for (stopName: String in sharedStops.keys) {

			// Try to remove stops by shared stop name.
			if (stops.remove(stopName) != null) {

				// If the value for removed wasn't null then log the stop for debugging.
				Log.d("purgeStops", "Stop removed: $stopName")
			}
		}
	}

	/**
	 * Sets the polylines visibly to whether or not the route is enabled or not.
	 * If the polyline didn't previously exist it will be created in this method.
	 *
	 * This must be run on the UI Thread.
	 *
	 * @param routeMatch The RouteMatch object to use if the polyline coordinates need to be retrieved.
	 * @param map        The map to add the polyline to.
	 */
	@UiThread
	fun togglePolylineVisibility(routeMatch: RouteMatch, map: GoogleMap) {

		if (polyline != null) {
			Log.d("togglePolyline", "Setting route $name to visible: $enabled")
			polyline!!.isVisible = enabled
			return
		}

		// We know the polyline is null here...
		// Get the polyline coordinates for the route from the RouteMatch server.
		routeMatch.callLandRoute(this,  {

			// Get the data from all the stops and store it in a JSONArray.
			val data: JSONArray = RouteMatch.parseData(it)

			// Get the land route points object from the land route data array.
			val landRoutePoints: JSONObject = data.getJSONObject(0)

			// Get the land route points array from the land route points object.
			val landRoutePointsArray: JSONArray = landRoutePoints.getJSONArray("points")

			// Get the number of points in the array.
			val count: Int = landRoutePointsArray.length()

			// Create a new LatLng array to store all the coordinates.
			val coordinates = arrayOfNulls<LatLng>(count)

			// Initialize the array of coordinates by iterating through the land route points array.
			for (i in 0 until count) {

				// Get the land route point object from the land route points array.
				val landRoutePoint = landRoutePointsArray.getJSONObject(i)

				// Get the latitude and longitude from the land route point.
				val latitude: Double = landRoutePoint.getDouble("latitude")
				val longitude: Double = landRoutePoint.getDouble("longitude")

				// Add the newly created LatLng object to the LatLng array.
				coordinates[i] = LatLng(latitude, longitude)
			}

			// Create a new polyline for the route since it didn't have one before.
			@Suppress("UNCHECKED_CAST") // Suppressed because we are asserting that none of the coordinates are null
			createPolyline(coordinates as Array<LatLng>, map)

			// Log if there was any error getting the polyline coordinates.
		} , { error: VolleyError -> Log.e("togglePolylineVisible", "Unable to get polyline coordinates", error) }, this)

	}

	/**
	 * Removes the routes polyline from the map, and sets it to null.
	 *
	 * This must be run on the UI thread.
	 */
	@UiThread
	fun removePolyline() {
		if (polyline != null) {
			polyline!!.remove()
			polyline = null
		}
	}

	companion object {

		/**
		 * Creates a new route object from the provided json object.
		 * If the json object is null then a RouteException will be thrown.
		 *
		 * @param jsonObject The json object contain the data to create a new route object.
		 * @return The newly created route object.
		 * @throws JSONException                Thrown if the json object is null,
		 *                                      or if the route name is unable to be parsed.
		 * @throws UnsupportedEncodingException Thrown if the route name cannot be formatted to a URL.
		 */
		@JvmStatic
		@Throws(JSONException::class, UnsupportedEncodingException::class)
		fun generateRoute(jsonObject: JSONObject): Route {

			// First, parse the name.
			val name = jsonObject.getString("routeId")

			return try {

				// Now try to parse the route color.
				val colorName = jsonObject.getString("routeColor")
				val color = Color.parseColor(colorName)

				// Return our newly created route with color!
				Route(name, color)
			} catch (exception: Exception) {
				Log.w("generateRoute", "Unable to parse color", exception)

				// Since there was an issue parsing the color,
				// and we have the name at this point simply create the route without a color.
				Route(name)
			}
		}

		/**
		 * Iterates though all the provided routes and enables those that have been marked favorited.
		 *
		 * @param routes             All the routes that can be tracked.
		 * @param favoriteRouteNames The names of the favorite routes.
		 */
		@JvmStatic
		fun enableFavoriteRoutes(routes: HashMap<String, Route>, favoriteRouteNames: Array<String>) {

			for (routeName: String in favoriteRouteNames) {
				try {
					routes[routeName]!!.enabled = true
				} catch (nullPointerException: NullPointerException) {

					// If the route was not found log it as an error.
					Log.w("enableFavoriteRoutes", "$routeName route is not running today", nullPointerException)
				}
			}
		}

		@JvmField
		val CREATOR = object : Parcelable.Creator<Route> {

			override fun createFromParcel(parcel: Parcel): Route { return Route(parcel) }

			override fun newArray(size: Int): Array<Route?> { return arrayOfNulls(size) }
		}
	}

	override fun describeContents(): Int { return this.hashCode() }

	override fun writeToParcel(parcel: Parcel, flags: Int) {
		parcel.writeString(this.name)
		parcel.writeInt(this.color)
		parcel.writeString(this.urlFormattedName)
		parcel.writeParcelableArray(this.stops.values.toTypedArray(), flags)
		parcel.writeParcelableArray(this.sharedStops.values.toTypedArray(), flags)
	}
}